composer/trainer/trainer.py

```diff
@@ -3655,7 +3665,19 @@ def _iter_dataloader(self, trainer_mode: TrainerMode):
                     # Otherwise, we will encounter an error at the start of the next epoch when
                     # Event.BEFORE_DATALOADER tries to start an unfinished marker.
                     self.engine.run_marker_only_event(Event.AFTER_DATALOADER)
+                # Mark iteration as finished - don't break yet as we need to sync across ranks
+                iter_finished += 1
+
+            # Sync iter finished across ranks
+            dist.all_reduce(iter_finished, reduce_operation='MAX')
+            # If any rank has finished, stop all rank iterations
+            if iter_finished.item() == 1:
                 break
+
+            if batch is sentinel:
+                raise RuntimeError(
+                    "Batch should have been assigned or loop should have been broken. This shouldn't happen!",
+                )
```

Comment:
I've added a unit test. It can be run with:
```
make test-dist-gpu WORLD_SIZE=2 PYTEST="pytest tests/trainer/test_trainer.py::TestTrainerInitOrFit::test_rank_dependent_dataloader_lengths"
```